### PR TITLE
BTS-735 Fix potential deadlocks

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

*(Please describe the changes in this PR for reviewers, motivation, rationale - **mandatory**)*

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/IResearch/IResearchLink.cpp [hunk 1]
```diff
@@ -1851,18 +1844,16 @@ Result IResearchLink::insert(transaction::Methods& trx,
 
   if (!ctx) {
     // '_dataStore' can be asynchronously modified
-    auto lock = _asyncSelf->lock();
-
-    if (!_asyncSelf.get()) {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchLink.cpp:1856]: bug here

File: arangod/IResearch/IResearchLink.cpp [hunk 2]
```diff
@@ -756,15 +745,14 @@ void IResearchLink::afterTruncate(TRI_voc_tick_t tick,
     THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
   }
 
-  if (!_asyncSelf) {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchLink.cpp:759]: Bug here

File: arangod/IResearch/IResearchLink.cpp [hunk 3]
```diff
@@ -879,33 +867,37 @@ Result IResearchLink::cleanupUnsafeImpl() {
 }
 
 Result IResearchLink::commit(bool wait /*= true*/) {
-  auto lock =
-      _asyncSelf->lock();  // '_dataStore' can be asynchronously modified
-
-  if (!_asyncSelf.get()) {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchLink.cpp:885]: Bug here

File: arangod/IResearch/IResearchView.h [hunk 4]
```diff
@@ -198,33 +199,35 @@ class IResearchView final : public arangodb::LogicalView {
...
-               // link did not come up yet) FIXME TODO maybe this should be
-               // asyncSelf?
+  AsyncViewPtr _asyncSelf;
+  // 'this' for the lifetime of the view (for use with asynchronous calls)
+  using Links = std::map<IndexId, AsyncLinkPtr>;
+  containers::FlatHashMap<DataSourceId, Links::iterator> _collections;
+  Links _links;
+  // (AsyncLinkPtr may be nullptr on single-server if link did not come up yet)
+  // map because we want same order for acquire locks (IResearchLink::snapshot)
+  // as PhysicalCollection::_indexes (IResearchLink::insert)
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.h:227]: Main issue

Comment (MBkkt) [inline arangod/IResearch/IResearchView.h:227]: Main issue

File: arangod/IResearch/IResearchView.cpp [hunk 5]
```diff
@@ -297,18 +282,21 @@ struct IResearchView::ViewFactory : public arangodb::ViewFactory {
...
-                                 // comes up it'll call link(...)
+      auto collection = vocbase.lookupCollection(cid);
+      // always look up in vocbase
+      // (single server or cluster per-shard collection)
+      auto link =
+          collection ? IResearchLinkHelper::find(*collection, *impl) : nullptr;
+      // add placeholders to links,
+      // when the collection comes up it'll bring up the link
+      auto [it1, was1] =
+          impl->_collections.try_emplace(cid, impl->_links.end());
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:293]: _collections same as old _links
_links only for collection link that is not null

File: arangod/IResearch/IResearchView.cpp [hunk 6]
```diff
@@ -599,175 +527,149 @@ Result IResearchView::dropImpl() {
...
   }
-
-  return linkPtr->propertiesUnsafe(_meta);
+  return IResearchLink::properties(std::move(linkLock), _meta);
 }
 
 Result IResearchView::commit() {
-  read_write_mutex::read_mutex mutex(
-      _mutex);  // '_links' can be asynchronously updated
+  // TODO Maybe we want sync all and skip errors, now we stop on first error?
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:648]: Check this plz

File: arangod/IResearch/IResearchView.cpp [hunk 7]
```diff
@@ -850,155 +741,139 @@ IResearchView::Snapshot const* IResearchView::snapshot(
...
-      _mutex);  // '_metaState' can be asynchronously modified
+  auto const& c = *collections;
+  read_write_mutex::read_mutex mutex(_mutex);
+  // '_metaState' can be asynchronously modified
   auto lock = irs::make_lock_guard(mutex);
-
   try {
     // collect snapshots from all requested links
-    for (auto const cid : *collections) {
+    for (auto& entry : _links) {
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:780]: your attention

File: arangod/IResearch/IResearchView.cpp [hunk 8]
```diff
@@ -850,155 +741,139 @@ IResearchView::Snapshot const* IResearchView::snapshot(
...
-      return Result();  // already unlinked
+    auto it = _collections.find(cid);
+    if (it == _collections.end()) {
+      return {};  // already unlinked
+    }
+    if (it->second != _links.end()) {
+      link = _links.extract(it->second);
     }
-
-    auto links = _links;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:967]: strange

File: arangod/IResearch/IResearchView.cpp [hunk 9]
```diff
@@ -850,155 +741,143 @@ IResearchView::Snapshot const* IResearchView::snapshot(
...
-
+      --maxFind;
+      auto linkLock = entry.second->lock();
+      if (!linkLock) {
+        continue;
+      }
+      auto cid = linkLock->collection().id();
+      if (!collections->contains(cid)) {
+        continue;
+      }
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:800]: We can check `collections->containes` before `->lock()` if we would store `collection id` in `entry`.
But I think `shared_lock` is not critical and this code ok

**Review by mpoeter — APPROVED** [review]

**Review by mpoeter — CHANGES_REQUESTED** [review]
If this PR fixes some of the issues currently listed in tsan_suppression, then these should be removed.

File: arangod/IResearch/IResearchView.cpp [hunk 10]
```diff
@@ -89,39 +83,46 @@ class ViewTrxState final : public arangodb::TransactionState::Cookie,
...
+    if (collections.size() != size) {
+      return false;
+    }  // TODO better iterate for smaller capacity
+    for (auto const& cid : _collections) {
+      if (!collections.contains(cid)) {
         return false;
       }
     }
-
-    return _collections.size() == count;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:103]: Don't know why we want equal but not greater subset

File: arangod/IResearch/IResearchLink.h [hunk 11]
```diff
@@ -126,8 +128,8 @@ class IResearchLink {
     }
 
    private:
-    AsyncValue<IResearchLink>::Value
-        _lock;  // lock preventing data store deallocation
+    AsyncValue<IResearchLink>::Value _lock;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchLink.h:131]: NIT
use `LinkLock`

File: arangod/IResearch/IResearchView.h [hunk 12]
```diff
@@ -198,33 +199,35 @@ class IResearchView final : public arangodb::LogicalView {
...
   IResearchViewMeta _meta;  // the view configuration
-  mutable arangodb::iresearch::kludge::read_write_mutex
-      _mutex;                   // for use with member '_meta', '_links'
+  mutable arangodb::iresearch::kludge::read_write_mutex _mutex;
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.h:229]: NIT
don't need to write `arangodb::iresearch::`

File: arangod/IResearch/IResearchView.cpp [hunk 13]
```diff
@@ -599,175 +531,146 @@ Result IResearchView::dropImpl() {
...
+  } else if (ServerState::instance()->isSingleServer() &&
+             it->second == _links.end()) {
+    auto [it1, was1] = _links.try_emplace(linkLock->id(), link);
+    TRI_ASSERT(was1);
+    it->second = it1;
+    IResearchLink::properties(std::move(linkLock), _meta);
+    // single-server persisted cid placeholder substituted with actual link
+    return {};
+  } else if (it->second != _links.end() &&
+             it->second->second->empty()) {  // TODO REPLACE
```
Comment (MBkkt) [inline arangod/IResearch/IResearchView.cpp:625]: remove todo

**Comment by MBkkt** [comment]
> If this PR fixes some of the issues currently listed in tsan_suppression, then these should be removed.
@mpoeter  You don't merge your PR in devel yet

**Comment by MBkkt** [comment]
https://github.com/arangodb/arangodb/pull/15576

**Pull request closed without merging**